### DISCUSSION TITLE
How to retrieve vim function arguments inside inline python code-blocks in Vim9?

### DISCUSSION THREAD

**Discussion by ubaldot** [vim/vim#15691] [Q&A]
Consider the following legacy Vim snippet

```
function! Test(foo)
  python3 << END
import vim

bar = vim.eval('a:foo')
print(bar)
END

endfunction

call Test('hello')
```

It works as expected. I guess the Vim9 counterpart should be something like the following:

```
vim9script

def Test(foo: string)
  python3 << END
import vim

bar = vim.eval('foo')
print(bar)
END

enddef

Test('hello')
```
but it ends up in a `E121: Undefined variable: foo`.
How shall I fix it? 

EDIT: as a workaround I am using the following:

```
vim9script

def Test(foo: string)
  g:foo  = foo
  python3 << END
import vim

bar = vim.eval('g:foo')
print(bar)
END
unlet g:foo
enddef

Test('hello')
```

**Comment by jessepav** [comment]
You can marshal your arguments through a script-local variable to avoid using the global namespace:

```
vim9script

var pyargs: list<any>

def Test(foo: string)
  pyargs = [foo]
  python3 << END
import vim

bar = vim.eval('pyargs[0]')
print(bar)
END
enddef

Test('hello')
```

**Comment by ubaldot** [comment]
That is also a nice approach, but I think it would be nicer if we can retrieve arguments directly, no? :)

**Comment by jessepav** [comment]
This issue was talked about back in 2021

https://github.com/vim/vim/issues/8573

If you don't want to use the script-local variable, you'll need to stuff your function arguments into the second parameter to `py3eval()`, as shown at the end of that thread.

**Reply by jessepav** [reply]
Oh wait, that `p3eval()` syntax was never implemented, so script-local is your only way to go.
